### PR TITLE
Role management

### PR DESCRIPTION
**Pull request by anbublacky**
## What does this change do?
- Add permission feature to the application

## Why is this change needed?
- Pundit policy can be made optional now. We can define policies from the application and set a scope for it.

## How were the changes done?
- Created Model for Role and Permission. Created CmAdmin Changes for the same.
- Created route to save the permission.
- Policies generated dynamically via the cm-admin, like how controller is generated.
- We can set scope for each action via the DSL for each model.

## How was testing done?
- Checked with geist project.
- Enabling an action and checking if the button gets disabled.
- Added a scope and checked if the scope is applied.

## Screenshots/Recording (optional)
image
image

### CODE REVIEW

File: app/views/cm_admin/roles/permissions.html.slim [hunk 1]
```diff
@@ -0,0 +1,33 @@
+.form-page.permission-form
+  .form-page__body
+    .form-container
+      = form_for Permission.new, url: "/cm_admin/roles/#{params[:id]}/create_role_permission", method: :post do |f|
+        - CmAdmin.config.cm_admin_models.each do |model|
+          - next if model.override_policy == true
+          .row
+            .col.form-container
+              p.form-title
+                = model.name.titleize
```
Comment (mikevic) [inline app/views/cm_admin/roles/permissions.html.slim:10]:
```suggestion
                = model.name.titleize
```

@anbublacky - we should default to the display_name of the model

File: lib/cm_admin/model.rb [hunk 2]
```diff
@@ -161,5 +173,60 @@ def user_not_authorized
...
+        end
+    
+        private
+    
+        attr_reader :user, :scope
+      end
+
+      policy.const_set 'Scope', klass
+
+      # klass = Class.new(policy) do
```
Comment (mikevic) [inline lib/cm_admin/model.rb:214]: @anbublacky - we should delete this?

**Review by mikevic — CHANGES_REQUESTED** [review]
@anbublacky 
1. Can we add docs on how to use this feature? It can be a separate page. 
2. What is the name of the task to add the migration for roles? 
3. Can we name the tables as cm_roles and cm_permissions? The main reason is to avoid clashing with any tables that are already present on the application. 
4. On the Geist PR, I saw some CSS that is added for this feature, can we add the CSS on CM Admin itself rather than on the project? 
5. Can you point me to where the default permission for admin is defined?

**Comment by anbublacky** [comment]
1. Can we add docs on how to use this feature? It can be a separate page.
   - .md file added
2. What is the name of the task to add the migration for roles?
   - rake cm_admin:install_role
3. Can we name the tables as cm_roles and cm_permissions? The main reason is to avoid clashing with any tables that are already present on the application.
   - Done
4. On the Geist PR, I saw some CSS that is added for this feature, can we add the CSS on CM Admin itself rather than on the project?
   - Migrated the CSS.
5. Can you point me to where the default permission for admin is defined?
  - rake task will create default migration. Linking the user to role to be manually done.

**Review by barath46 — APPROVED** [review]

**Pull request merged**